scripts/util/doc-builder.js

```diff
@@ -0,0 +1,341 @@
+      };
+
+      self.log('info', 'Checking out the tree to ', options.targetDirectory);
+      return NodeGit.Checkout.tree(repo, tree, options);
+    }).then(function () {
+      self.log('info', 'Done checking out ', self.options.docsBranchName);
+      if (callback) {
+        callback();
+      }
+    });
+  }).catch(function (error) {
+    self.log('info', 'Error checking out Documentations Branch');
+    console.error(error);
+    if (callback) {
+      callback(error);
+    }
+  });
+}
+
+DocBuilder.prototype.checkoutAPISrc = function (callback) {
+  var self = this;
+  if (!fs.existsSync(self.options.apiDirPath)) {
+    fs.mkdirSync(self.options.apiDirPath);
+  }
+
+  self.log('info', '\nChecking out source code for each version tag and the ', self.options.masterBranchName, ' branch');
+  NodeGit.Repository.open('.').then(function (repo) {
+
+    self.log('info', 'Getting available references within the repo');
+    return repo.getReferences(NodeGit.Reference.TYPE.OID).then(function(references){
+      var treesDataPromises = [];
+
+      self.log('info', 'Filtering them to include only tags and the ', self.options.masterBranchName, ' branch.');
+      references.forEach(function(reference){
+        var target = reference.targetPeel() || reference.target();
+        var name = reference.shorthand();
+        var commitPromise;
+
+        if(reference.isTag()) {
+          commitPromise = repo.getCommit(target);
+        } else if (reference.isBranch() && reference.shorthand() === self.options.masterBranchName) {
+          commitPromise = repo.getBranchCommit(name);
+        } else {
+          return;
+        }
+        self.log('info', 'Gotten a promise for commit for ', name);
+        self.log('info', 'Adding a promise to get commit tree for ', name);
+        treesDataPromises.push(buildTreePromise(commitPromise, name));
+      });
+
+      self.log('info', 'Resolving commit tree promises');
+      return Promise.all(treesDataPromises);
+    }).then(function(treesData){
+      var paths = [];
+      var options = JSON.parse(JSON.stringify(self.options));
+      options.repo = repo;
+
+      self.log('info', 'Checking out commit trees');
+      checkoutTrees(treesData.shift(), treesData, options, function (paths) {
+        self.log('info', 'Done checking out source code');
+
+        if (callback) {
+          callback(null, paths);
+        }
+      });
+    });
+  }).catch(function(error){
+    console.error("Error building doc");
+    console.error(error);
+    if (callback) {
+      callback(error);
+    }
+  });
+}
+
+DocBuilder.prototype.generateAPIDocs = function (paths, callback) {
+  var self = this;
+
+  self.log('info', '\nGenerating API docs');
+  yuidoc.generate({paths: paths, themeDir: self.options.themeDir}, function () {
+    self.log('info', 'Done generating API docs');
+    if (callback) {
+      callback();
+    }
+  });
+}
+
+DocBuilder.prototype.commitAPIDocs = function (callback) {
```

Comment:
these mixed in callbacks w/Promises add a lot of complexity.  Can these methods return promises instead?